pkg/storage/snapshot/restore_test.go

```diff
@@ -415,21 +422,54 @@ var _ = Describe("Restore controller", func() {
 				storageClassSource.Add(storageClass)
 			})
 
-			It("should error if snapshot does not exist", func() {
+			DescribeTable("should error if snapshot", func(vmSnapshot *snapshotv1.VirtualMachineSnapshot, expectedError string) {
 				r := createRestoreWithOwner()
 				vm := createModifiedVM()
 				rc := r.DeepCopy()
 				rc.ResourceVersion = "1"
 				rc.Status = &snapshotv1.VirtualMachineRestoreStatus{
 					Complete: pointer.P(false),
 					Conditions: []snapshotv1.Condition{
-						newProgressingCondition(corev1.ConditionFalse, "VMSnapshot default/snapshot does not exist"),
-						newReadyCondition(corev1.ConditionFalse, "VMSnapshot default/snapshot does not exist"),
+						newProgressingCondition(corev1.ConditionFalse, expectedError),
+						newReadyCondition(corev1.ConditionFalse, expectedError),
 					},
 				}
 				vmSnapshotSource.Delete(createSnapshot())
+				if vmSnapshot != nil {
+					vmSnapshotSource.Add(vmSnapshot)
+				}
 				vmSource.Add(vm)
-				expectUpdateVMRestoreInProgress(vm)
+				updateStatusCalls := expectVMRestoreUpdateStatus(kubevirtClient, rc)
+				addVirtualMachineRestore(r)
+				controller.processVMRestoreWorkItem()
+				testutils.ExpectEvent(recorder, "VirtualMachineRestoreError")
+				Expect(*updateStatusCalls).To(Equal(1))
+			},
+				Entry("does not exist", nil, "VMSnapshot default/snapshot does not exist"),
+				Entry("in failed state", createSnapshotWith(snapshotv1.Failed, false), "VMSnapshot default/snapshot failed and is invalid to use"),
+				Entry("not ready", createSnapshotWith(snapshotv1.InProgress, false), "VMSnapshot default/snapshot not ready"),
+			)
+
+			It("should error if target exists before the restore and it is not the same as the source", func() {
+				r := createRestoreWithOwner()
+				vm := createModifiedVM()
+				newVM := createVirtualMachine(testNamespace, newVMName)
+				newVM.UID = newVMUID
+				r.Spec.Target.Name = newVM.Name
+
+				rc := r.DeepCopy()
+				rc.ResourceVersion = "1"
+				rc.Status = &snapshotv1.VirtualMachineRestoreStatus{
+					Complete: pointer.P(false),
+					Conditions: []snapshotv1.Condition{
+						newProgressingCondition(corev1.ConditionFalse, "restore source and restore target are different but restore target already exists"),
+						newReadyCondition(corev1.ConditionFalse, "restore source and restore target are different but restore target already exists"),
```

Comment:
sure you want to repeat the same reason in both?

Reply:
thats whats done for other condition changes.. What would you suggest?

Reply:
to me, it seems more relevant to the progressing condition, but I see why you have them repeated https://github.com/kubevirt/kubevirt/blob/dcdef7be5ea7084fc3d9817fe7ce5ca8f1e60c3e/pkg/storage/snapshot/restore.go#L269-L270 so up to you